Add support for isFabricScoped on a command & use it in TLSCertificateManagementCluster

gmarcosb opened the pull request:
Add support for commands marked with `isFabricScoped`

The behavior of these commands is specified at decode time if `source=client`, with the following method signature:
```
CHIP_ERROR DecodableType::Decode(TLV::TLVReader &reader, FabricIndex aAccessingFabricIndex)
```

And at encode time if `source=server`, with the following updated method signature:
```
CHIP_ERROR Type::Encode(TLV::TLVWriter & aWriter, TLV::Tag aTag, FabricIndex aAccessingFabricIndex) const {
```

This is to guarantee consistent behavior on the receiver of the command (whether client or server), such that the fabric index is always set to by the server.

This is to match attribute behavior.

#### Testing
Verified by CI

gmarcosb committed:
Add support for isFabricScoped on command

gmarcosb committed:
Re-introduce isFabricScoped on structs & commands

gmarcosb committed:
Generated using ./scripts/tools/zap_regen_all.py

bzbarsky-apple commented on src/app/zap-templates/templates/app/cluster-objects.zapt:
```diff
@@ -108,12 +108,21 @@ public:
     {{zapTypeToEncodableClusterObjectType type}} {{asLowerCamelCase label}}{{> cluster_objects_field_init}};
     {{/zcl_command_arguments}}
 
+    {{#if isFabricScoped~}}
+    {{#if (isServer source)}}
```
What does it even mean for a server-source command to be fabric-scoped?  fabric-scoped in the spec for a command means "command must be associated with a fabric on the recipient of the command", no?   Is it actually used for any server-source commands, and what does it mean for those?

gmarcosb replied:
🤷‍♂️ so in this PR, if you look at `TLSCertStruct`, it is marked as `isFabricScoped`, which is then used in `FindRootCertificateResponse`

I mean if an attribute can be fabric-scoped, I suppose it makes sense that a server-source command can too, right? (aren't they effectively the same, except attributes are pull & server commands are push?)

Here is that fabric scoped struct in spec: https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/f2136a2784837b93efb17814a57f2d4ef29d836a/src/tls/TLSCertificateManagement.adoc#ref_TLSCertStruct

I will defer to you here, I'm just trying to get the spec compiling 🤣

bzbarsky-apple commented on src/app/zap-templates/templates/app/cluster-objects.zapt:
```diff
@@ -129,11 +138,21 @@ struct DecodableType {
     {{zapTypeToDecodableClusterObjectType type}} {{asLowerCamelCase label}}{{> cluster_objects_field_init}};
     {{/zcl_command_arguments}}
+
+    {{~#if isFabricScoped~}}
+    {{#if (isClient source)}}
+    CHIP_ERROR Decode(TLV::TLVReader &reader, FabricIndex aAccessingFabricIndex);
```
The relevant thing here, per Slack discussion, is whether the command has structs that are fabric-scoped as command fields, not whether the command itself is fabric-scoped, right?  Does anything in the spec actually require a command with a fabric-scoped struct field to itself be fabric-scoped?

gmarcosb replied:
This is really more of a code requirement than a spec requirement - i.e. the command is the root, and so we need a way (at the root) to pass the fabric index to the underlying structs

What we could do (an alternative) would simply be to change all `Command::Decode` to always take in a `aAccessingFabricIndex`

I figured the advantage of the approach I've taken here is that it's more explicit, i.e. the caller will be aware that the command has data which is fabric scoped by having to pass in a fabric index

(We could imply from the sub-structs, but at that point we are effectively implicitly marking the command as fabric scoped, so seems reasonable to make it explicit)

bzbarsky-apple commented on src/app/data-model/Decode.h:
```diff
@@ -114,6 +114,27 @@ CHIP_ERROR Decode(TLV::TLVReader & reader, X & x)
+ * @brief
+ *
+ * This specific variant that decodes cluster objects (like structs, commands, events) from TLV
+ * depends on the presence of a Decode method on the object to present. The signature of that method
```
I know this got copy/pasted, but what does "to present" mean here?  Looks like an editing error of some sort in the comment this got copied from....

gmarcosb replied:
Updated to template name (X)

bzbarsky-apple commented on src/app/zap-templates/templates/app/im-cluster-command-handler.zapt:
```diff
@@ -37,7 +37,12 @@ Protocols::InteractionModel::Status DispatchServerCommand(CommandHandler * apCom
     {{/first}}
         case Commands::{{asUpperCamelCase commandName}}::Id: {
             Commands::{{asUpperCamelCase commandName}}::DecodableType commandData;
+
```
```suggestion
```
I see no reason for the whitespace churn there...

gmarcosb replied:
Woops, apologies, I looked at the output of cluster-objects-src.zapt so carefully but forgot to double-check this one

bzbarsky-apple commented on src/app/zap-templates/templates/app/im-cluster-command-handler.zapt:
```diff
@@ -37,7 +37,12 @@ Protocols::InteractionModel::Status DispatchServerCommand(CommandHandler * apCom
             Commands::{{asUpperCamelCase commandName}}::DecodableType commandData;
+
+            {{#if isFabricScoped~}}
+            TLVError = DataModel::Decode(aDataTlv, commandData, apCommandObj->GetAccessingFabricIndex());
```
I have to ask: do we even need the new DataModel::Decode overload?  Why can't we just `commandData.Decode(...)` here?

gmarcosb replied:
I tried to minimize how much generated code we have 😅🤣

I.e. if it could be pulled out to a common method instead of being generated for every command, I did so

gmarcosb replied:
Took a slightly different approach here, adding an overload to Decode method so we don't have to check `isFabricScoped` here; let me know what you think (i.e. if you prefer calling `commandDecode`, will make the change

gmarcosb committed:
Apply code review suggestions

gmarcosb commented on examples/chip-tool/templates/logging/DataModelLogger-src.zapt:
```diff
@@ -122,7 +124,7 @@ CHIP_ERROR DataModelLogger::LogCommand(const chip::app::ConcreteCommandPath & pa
             {
                 {{asUpperCamelCase parent.name}}::Commands::{{asUpperCamelCase name}}::DecodableType value;
-                ReturnErrorOnFailure(chip::app::DataModel::Decode(*data, value));
+                ReturnErrorOnFailure(chip::app::DataModel::Decode(*data, value, kUndefinedFabricIndex));
```
This is less than ideal... but not sure what else we can do (`LogAttribute` similarly will log bad data; it's just not obvious because it doesn't have the new method override like commands do)

gmarcosb committed:
Improve calls to Decode

bzbarsky-apple commented on zzz_generated/app-common/app-common/zap-generated/cluster-objects.cpp:
```diff
@@ -948,7 +940,7 @@ CHIP_ERROR Type::Encode(TLV::TLVWriter & aWriter, TLV::Tag aTag) const
 }
 
-CHIP_ERROR DecodableType::Decode(TLV::TLVReader & reader)
+CHIP_ERROR DecodableType::Decode(TLV::TLVReader & reader, FabricIndex aAccessingFabricIndex)
```
Many many commands are fabric-scoped that have nothing to do with structs at all.  All fabric-scoped means for a command is "must be associated with a fabric".

gmarcosb replied:
OK cool, well then this should be fine 👍🙏

bzbarsky-apple commented on examples/chip-tool/templates/logging/DataModelLogger-src.zapt:
```diff
@@ -122,7 +124,7 @@ CHIP_ERROR DataModelLogger::LogCommand(const chip::app::ConcreteCommandPath & pa
             {
                 {{asUpperCamelCase parent.name}}::Commands::{{asUpperCamelCase name}}::DecodableType value;
-                ReturnErrorOnFailure(chip::app::DataModel::Decode(*data, value));
+                ReturnErrorOnFailure(chip::app::DataModel::Decode(*data, value, kUndefinedFabricIndex));
```
Is this logging incoming or outgoing commands?

gmarcosb replied:
🤷‍♂️😬 how do I find out?

gmarcosb replied:
Does seem to be client-side & unneeded after last change

gmarcosb committed:
Revert unnecessary change from client-side changes

gmarcosb committed:
Merge conflicts from https://github.com/project-chip/connectedhomeip/pull/38390

gmarcosb committed:
Additional changes to support EncodeForRead with accessingFabricIndex

gmarcosb committed:
Add kIsFabricScoped to hand-rolled structs

gmarcosb commented on src/app/CommandHandlerImpl.h:
```diff
@@ -436,7 +436,7 @@ class CommandHandlerImpl : public CommandHandler
      */
     bool IsGroupRequest() { return mGroupRequest; }
 
-    bool ResponsesAccepted() { return !(mGroupRequest || mpResponder == nullptr); }
```
This change was unrelated, but made it more readable to reason about when responses are accepted (i.e. mResponder must not be null), vs the previous conditional which negated the whole thing

tcarmelveilleux committed:
Add VID Verification Statement to TC-RR-1.1 (#38712)

* Add VID Verification Statement to TC-RR-1.1

- Add VID Verification Statement filling to TC-RR-1.1
  - Matches https://github.com/CHIP-Specifications/chip-test-plans/pull/5131
- Add missing logging
- Add missing checks as well that were listed in test plan but not actually
  implemented.
- Add native methods missing to implement test changes

* Restyled by clang-format

* Restyled by autopep8

* Fix lint

* Update src/python_testing/TC_RR_1_1.py

Co-authored-by: C Freeman <cecille@google.com>

* Fix value access in attestation challenge getter

---------

Co-authored-by: Restyled.io <commits@restyled.io>
Co-authored-by: C Freeman <cecille@google.com>

tcarmelveilleux committed:
Add VID Verification Statement to TC-RR-1.1 (#38712)

* Add VID Verification Statement to TC-RR-1.1

- Add VID Verification Statement filling to TC-RR-1.1
  - Matches https://github.com/CHIP-Specifications/chip-test-plans/pull/5131
- Add missing logging
- Add missing checks as well that were listed in test plan but not actually
  implemented.
- Add native methods missing to implement test changes

* Restyled by clang-format

* Restyled by autopep8

* Fix lint

* Update src/python_testing/TC_RR_1_1.py

Co-authored-by: C Freeman <cecille@google.com>

* Fix value access in attestation challenge getter

---------

Co-authored-by: Restyled.io <commits@restyled.io>
Co-authored-by: C Freeman <cecille@google.com>

gmarcosb committed:
Generated using ./scripts/tools/zap_regen_all.py

pidarped commented on src/app/CommandHandlerImpl.cpp:
```diff
@@ -135,7 +135,16 @@ CHIP_ERROR CommandHandlerImpl::TryAddResponseData(const ConcreteCommandPath & aR
+        ReturnErrorOnFailure(aEncodable.EncodeTo(*writer, TLV::ContextTag(CommandDataIB::Tag::kFields),
+                                                 context->GetSessionHandle()->GetFabricIndex()));
+    }
+    else
```
This else block is when the Exchange does not have a session at all, right? Would it still need to Encode? Should this be an outright failure?

gmarcosb replied:
I have no idea 😬🤦‍♂️

It seemed like the infra knowing what fabric we're encoding data for is something we should always have available; but without this check, some tests were crashing

I don't understand why CommandHandlerImpl doesn't always have the fabric the command is scoped to.. filed https://github.com/project-chip/connectedhomeip/issues/39001

sxb427 reviewed (commented):
/pullapprove-test

```yaml
version: 3

# https://developer.github.com/v3/previews/#draft-pull-requests
github_api_version: "shadow-cat-preview"

############################################################
#  Overrides
############################################################

overrides:
  - if: "'hotfix' in labels"
    status: success
    explanation: "Hotfix label added, bypassing reviews"

  ############################################################
  #  Draft PRs
  ############################################################
  - if: "draft"
    status: pending
    explanation: "PR is draft, pending review"

  ############################################################
  #  License Checks
  ############################################################
  - if: "'*license/cla*' not in statuses.successful"
    status: pending
    explanation: "CLA must be agreed to by all contributors"

  ############################################################
  #  Conditions to Skip Review
  ############################################################
  - if: "base.ref != 'master'"
    status: success
    explanation: "Review not required unless merging to master"

  ############################################################
  #  Require Issues
  ############################################################
  # disabling until we have PRs up to date
  #     - if: "'*issue*' not in statuses.successful"
  #       status: failure
  #       explanation: "An issue is required for all PRs"

  ############################################################
  #  Fast tracking
  ############################################################
  - if: "'fast track' in labels"
    status: success
    explanation: "PR has been fast tracked, bypassing reviews"

############################################################
#  Notifications
############################################################

notifications:
  ############################################################
  #  New contributors
  ############################################################
  - when: pull_request.opened
    if: "author_association == 'FIRST_TIME_CONTRIBUTOR'"
    comment: |
      Hey @{{ author }}, thanks for the PR! The review will start once
      the tests and CI checks have passed. If they don't, please review
      the logs and try to fix the issues (ask for help if you can't
      figure it out). A reviewer will be assigned once the tests are
      passing and they'll walk you through getting the PR finished
      and merged.

groups:
  ############################################################
  #  Shared Reviewer Groups
  ############################################################
  shared-reviewers-amazon:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-amazon]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-apple:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-apple]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-bosch:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-bosch]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-comcast:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-comcast]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-dyson:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-dyson]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-espressif:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-espressif]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-google:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-google]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-grundfos:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-grundfos]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-irobot:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-irobot]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-lg:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-lg]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-logitech:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-logitech]
    reviews:
      request: 0 # Requested to be only on demand
  shared-reviewers-nordic:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-nordic]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-nxp:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-nxp]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-samsung:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-samsung]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-eve:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-eve]
    reviews:
      request: 0 # Do not auto-add
  # shared-reviewers-signify disabled for now, because the reviewers-signify
  # team is empty and pullapprove seems to mis-handle that badly and treats
  # _all_ reviewers as being in this group.
  #
  # See https://github.com/dropseed/pullapprove/issues/71
  #
  # shared-reviewers-signify:
  #     type: optional
  #     conditions:
  #         - files.include('*')
  #     reviewers:
  #         teams: [reviewers-signify]
  #     reviews:
  #         request: 0 # Do not auto-add
  shared-reviewers-silabs:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-silabs]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-somfy:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-somfy]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-tcl:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-tcl]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-qorvo:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-qorvo]
    reviews:
      request: 0 # Do not auto-add

  ############################################################
  #  Path specific review requirments. A PR must meet the
  #  review requirements corresponding each path in the
  #  in the CUSTOM_REVIEW_PATHS list that it touches. If it
  #  touches a region not covered by any custom review paths, 
  #  base condition is required.
  ############################################################

  ############################################################
  #  Review requirements for PRs touching scripts/build*.
  #  One shared-reviewers-scripts-build review is sufficient.
  ############################################################
  folder-reviewers-scripts-build:
    type: optional
    conditions:
      - "files.include('scripts/build/*') or files.include('integrations/docker/*')"
    reviewers:
      users:
        - andy31415
    reviews:
      # 1 review request will be sent at a time
      request: 1

  ############################################################
  #  Review requirements for PRs touching examples/chef*
  #  One Googler requirement is sufficient.
  ############################################################
  folder-reviewers-google:
    type: required
    conditions:
      - "files.include('examples/chef/*') or files.include('integrations/cloudbuild/*')"
    requirements:
      - "len(groups.approved.include('shared-reviewers-google')) >= 1 or \
        len(groups.approved.include('shared-reviewers-*')) >= 2"
    reviews:
      required: 0

  ############################################################
  #  Review requirements for PRs touching src/darwin
  #  One Apple review is required.
  ############################################################
  folder-reviewers-src-darwin:
    type: required
    conditions:
      - "files.include('src/darwin/*').exclude('src/darwin/Framework/CHIP/zap-generated/*') or \
        files.include('src/platform/Darwin/*')"
    reviewers:
      teams: [reviewers-apple]
    reviews:
      request: 0
      required: 1
    labels:
      approved: "[Darwin] review - approved"
      pending: "[Darwin] review - pending"
      rejected: "[Darwin] review - changed requested"

  ############################################################
  #  Base Required Reviewers
  ############################################################
  required-reviewers:
    description: >
      [Required
      Reviewers](https://github.com/project-chip/connectedhomeip/blob/master/CONTRIBUTING.md#review-requirements)
      This is the main group of required reviews for general pull
      requests.
    type: required
    conditions:
      # If PR touches a region not covered by any custom review paths, base condition is required.
      - "files\
              .exclude('examples/chef/*')\
              .exclude('integrations/cloudbuild/*')\
              .exclude('src/darwin/*')\
              .exclude('src/platform/Darwin/*') or \
        files.include('src/darwin/Framework/CHIP/zap-generated/*')"
    requirements:
      - len(groups.approved.include('shared-reviewers-*')) >= 2
    reviews:
      required: 0
    labels:
      approved: "review - approved"
      pending: "review - pending"
      rejected: "review - changed requested"
```

sxb427 reviewed (commented):
/pullapprove-test

```yaml
version: 3
groups:
version: 3

# https://developer.github.com/v3/previews/#draft-pull-requests
github_api_version: "shadow-cat-preview"

############################################################
#  Overrides
############################################################

overrides:
  - if: "'hotfix' in labels"
    status: success
    explanation: "Hotfix label added, bypassing reviews"

  ############################################################
  #  Draft PRs
  ############################################################
  - if: "draft"
    status: pending
    explanation: "PR is draft, pending review"

  ############################################################
  #  License Checks
  ############################################################
  - if: "'*license/cla*' not in statuses.successful"
    status: pending
    explanation: "CLA must be agreed to by all contributors"

  ############################################################
  #  Conditions to Skip Review
  ############################################################
  - if: "base.ref != 'master'"
    status: success
    explanation: "Review not required unless merging to master"

  ############################################################
  #  Require Issues
  ############################################################
  # disabling until we have PRs up to date
  #     - if: "'*issue*' not in statuses.successful"
  #       status: failure
  #       explanation: "An issue is required for all PRs"

  ############################################################
  #  Fast tracking
  ############################################################
  - if: "'fast track' in labels"
    status: success
    explanation: "PR has been fast tracked, bypassing reviews"

############################################################
#  Notifications
############################################################

notifications:
  ############################################################
  #  New contributors
  ############################################################
  - when: pull_request.opened
    if: "author_association == 'FIRST_TIME_CONTRIBUTOR'"
    comment: |
      Hey @{{ author }}, thanks for the PR! The review will start once
      the tests and CI checks have passed. If they don't, please review
      the logs and try to fix the issues (ask for help if you can't
      figure it out). A reviewer will be assigned once the tests are
      passing and they'll walk you through getting the PR finished
      and merged.

groups:
  ############################################################
  #  Shared Reviewer Groups
  ############################################################
  shared-reviewers-amazon:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-amazon]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-apple:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-apple]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-bosch:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-bosch]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-comcast:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-comcast]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-dyson:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-dyson]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-espressif:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-espressif]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-google:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-google]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-grundfos:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-grundfos]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-irobot:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-irobot]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-lg:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-lg]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-logitech:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-logitech]
    reviews:
      request: 0 # Requested to be only on demand
  shared-reviewers-nordic:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-nordic]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-nxp:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-nxp]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-samsung:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-samsung]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-eve:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-eve]
    reviews:
      request: 0 # Do not auto-add
  # shared-reviewers-signify disabled for now, because the reviewers-signify
  # team is empty and pullapprove seems to mis-handle that badly and treats
  # _all_ reviewers as being in this group.
  #
  # See https://github.com/dropseed/pullapprove/issues/71
  #
  # shared-reviewers-signify:
  #     type: optional
  #     conditions:
  #         - files.include('*')
  #     reviewers:
  #         teams: [reviewers-signify]
  #     reviews:
  #         request: 0 # Do not auto-add
  shared-reviewers-silabs:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-silabs]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-somfy:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-somfy]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-tcl:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-tcl]
    reviews:
      request: 0 # Do not auto-add
  shared-reviewers-qorvo:
    type: optional
    conditions:
      - files.include('*')
    reviewers:
      teams: [reviewers-qorvo]
    reviews:
      request: 0 # Do not auto-add

  ############################################################
  #  Folder specific review requirements. A PR must meet all
  #  folder specific review requirements for all folders that
  #  it touches with specific requirements defined.
  ############################################################

  folder-reviewers-scripts-build-integrations-docker:
    type: optional
    conditions:
      - "files.include('scripts/build/*') or files.include('integrations/docker/*')"
    reviewers:
      users:
        - andy31415
    reviews:
      # 1 review request will be sent at a time
      request: 1

  folder-reviewers-google:
    type: required
    conditions:
      - "files.include('examples/chef/*') or files.include('integrations/cloudbuild/*')"
    requirements:
      - "len(groups.approved.include('shared-reviewers-google')) >= 1 or \
        len(groups.approved.include('shared-reviewers-*')) >= 2"
    reviews:
      required: 0

  folder-reviewers-apple:
    type: required
    conditions:
      - "files.include('src/darwin/*').exclude('src/darwin/Framework/CHIP/zap-generated/*') or \
        files.include('src/platform/Darwin/*')"
    reviewers:
      teams: [reviewers-apple]
    reviews:
      request: 0
      required: 1

  ############################################################
  #  Base Required Reviewers. If PR touches a path not covered
  #  by any folder specific review requirements.
  ############################################################
  required-reviewers:
    description: >
      [Required
      Reviewers](https://github.com/project-chip/connectedhomeip/blob/master/CONTRIBUTING.md#review-requirements)
      This is the main group of required reviews for general pull
      requests.
    type: required
    conditions:
      # If PR touches a region not covered by any custom review paths, base condition is required.
      - "files\
              .exclude('examples/chef/*')\
              .exclude('integrations/cloudbuild/*')\
              .exclude('src/darwin/*')\
              .exclude('src/platform/Darwin/*') or \
        files.include('src/darwin/Framework/CHIP/zap-generated/*')"
    requirements:
      - len(groups.approved.include('shared-reviewers-*')) >= 2
    reviews:
      required: 0
    labels:
      approved: "review - approved"
      pending: "review - pending"
      rejected: "review - changed requested"
```
